synapse/handlers/sync.py

```diff
@@ -661,16 +661,12 @@ async def get_state_at(
             stream_position: point at which to get state
             state_filter: The state filter used to fetch state from the database.
         """
-        # FIXME this claims to get the state at a stream position, but
-        # get_recent_events_for_room operates by topo ordering. This therefore
-        # does not reliably give you the state at the given stream position.
-        # (https://github.com/matrix-org/synapse/issues/3305)
-        last_events, _ = await self.store.get_recent_events_for_room(
-            room_id, end_token=stream_position.room_key, limit=1
+        last_event = await self.store.get_last_event_in_room_before_stream_ordering(
```

Comment:
```suggestion
        # FIXME: This gets the state at the latest event before the stream ordering,
        # which might not be the same as the "current state" of the room at the time
        # of the stream token if there were multiple forward extremities at the time.
        last_event = await self.store.get_last_event_in_room_before_stream_ordering(
```

This PR makes things a lot better, but its still not technically 100% correct. I think it makes sense to then leave in a `FIXME` comment.